Fix the back end search results handling

zoglo opened the pull request:
### Description

* Fixes a regression caused by #8019 and allows clicking search targets again
* Also adds a focus lock if search results exist

zoglo committed:
Do not close the search if a search result is clicked

zoglo committed:
CS

leofeyer reviewed (changes requested):
Doesn't work for me in Safari on macOS 😞

leofeyer commented:
It does work in Firefox though.

zoglo committed:
Use setTimeout and document.activeElement instead

zoglo committed:
Use existing states to handle `inputBlur`

zoglo commented:
@leofeyer a274afed749b0eac8275b0e8de1d3e7bceb397f2 uses the existing states now (they were already introcuded for the classes), thus preventing closing when we do have search results.

leofeyer commented:
I can click on the results now, yes, but tabbing is still broken, I'm afraid.

### Problem 1: no focus lock

https://github.com/user-attachments/assets/11e24f1b-fbfd-4bec-bebc-af0b259d2776

The focus should be locked until the search menu is closed with ESC.

### Problem 2: search results unreachable

https://github.com/user-attachments/assets/488c75b0-1099-4888-872e-7eb8bcd738f0

After entering a search term that yields results, it should be possible to tab through them. Oddly enough, this works when I search for `demo` but not if I search for `result` 🤔 But even then, it is possible to continue tabbing out of the search.

zoglo committed:
Merge branch '5.5' into fix/search-related-target-blur

zoglo committed:
Add a `focusTrap` when search is active

zoglo committed:
Expand the hit target and do not overflow long serialized results

zoglo committed:
Add custom focus styles (safari sometimes does not render the normal styles)

zoglo commented:
> **Problem 1: no focus lock**

Added in 61dc7d187b56be7bc98d648908af3e9a08fc0f9a

> **Problem 2: search results unreachable**

They are actually reachable but safari does not render the focus sometimes (graphical bug).
I added custom focus styles now ad3a8cdb79bd8bc3d685ae432483d990fa01634e
![image](https://github.com/user-attachments/assets/d77dedb6-2ce3-4705-9221-b5c3bf0cd2e5)
![image](https://github.com/user-attachments/assets/e47019e4-c6b7-48db-bf29-27eb36ff5bf0)
![image](https://github.com/user-attachments/assets/845c228a-fdfa-42a5-b621-8fdfde63b0f5)
![image](https://github.com/user-attachments/assets/54918281-dcb8-445a-af7b-c0210ebdb290)

zoglo commented:
@leofeyer ready to review once again

leofeyer committed:
Merge branch '5.5' into fix/search-related-target-blur

leofeyer committed:
Rebuild the assets

leofeyer commented:
### Existing bugs

* Bug 1: clicking search targets → fixed ✅ 
* Bug 2: no focus lock → fixed ✅ 
* Bug 3: search results unreachable → fixed ✅ 

### New bugs 🙈 

Bug 4: custom focus styles → I don't think it is a good idea to use different focus styles than in the rest of the application. The UI should be consistent and predictable. So since this is a Safari bug, I'd suggest we revert the custom focus style, investigate a little more and then discuss if we want to change the focus style application-wide. WDYT?

Bug 5: the tabbing order is still broken → open the back end and start tabbing until the search menu opens. Then press `ESC` to close the search bar → the tab focus resets and you have to start tabbing at the top again.

zoglo committed:
Remove the custom focus styles

zoglo commented:
> Bug 4: custom focus styles → I don't think it is a good idea to use different focus styles than in the rest of the application. The UI should be consistent and predictable. So since this is a Safari bug, I'd suggest we revert the custom focus style, investigate a little more and then discuss if we want to change the focus style application-wide. WDYT?

Yeah, we should - Removed in 1c346e1e437c24f5d1282e67b311cc4eb2d45898

zoglo committed:
Do not fire the `inputBlur` event

zoglo commented:
> Bug 5: the tabbing order is still broken → open the back end and start tabbing until the search menu opens. Then press `ESC` to close the search bar → the tab focus resets and you have to start tabbing at the top again.

@leofeyer Feels like a safari but on the blur event as well I guess but you did remove a lot of input blur events due to that:

* fixed in ~~2a961d53449f03ae3f6d2b196ef59ee758ab2bc9~~

Has been reverted since that is most likely a bug in safari :/

zoglo committed:
Merge branch '5.5' into fix/search-related-target-blur

m-vo commented on core-bundle/assets/controllers/backend-search-controller.js:
```diff
@@ -74,13 +74,47 @@ export default class BackendSearchController extends Controller {
+
+    focusTrapNext(event) {
+        if ("results" === this.state && document.activeElement === this.lastFocus) {
+            event.preventDefault();
```
```suggestion
```

m-vo commented on core-bundle/assets/controllers/backend-search-controller.js:
```diff
@@ -74,13 +74,47 @@ export default class BackendSearchController extends Controller {
+
+    focusTrapPrev(event) {
+        if ("results" === this.state && document.activeElement === this.firstFocus) {
+            event.preventDefault();
```
```suggestion
```

fritzmg replied:
Is this correct? With `:prevent` in the markup `event.preventDefault()` will _always_ be executed - and not conditionally like in this case.

m-vo commented on core-bundle/contao/templates/twig/backend/search/_container.html.twig:
```diff
@@ -3,7 +3,12 @@
+    .set('data-action','
+        click@document->contao--backend-search#documentClick
+        keydown.esc@document->contao--backend-search#close
+        keydown.tab@document->contao--backend-search#focusTrapNext
+        keydown.shift+tab@document->contao--backend-search#focusTrapPrev
```
```suggestion
        keydown.tab@document->contao--backend-search#focusTrapNext:prevent
        keydown.shift+tab@document->contao--backend-search#focusTrapPrev:prevent
```

zoglo committed:
Use the `prevent` flag instead

Co-authored-by: M. Vondano <m-vo@users.noreply.github.com>

zoglo committed:
Revert 70058a01af12e4d52600596e8cd519907d793268

zoglo committed:
Revert 2a961d53449f03ae3f6d2b196ef59ee758ab2bc9

zoglo committed:
Rebuild the assets

leofeyer committed:
Rebuild the assets

leofeyer commented:
Thank you @zoglo.
